File: backend/quiz-feedback-service/src/main/java/com/elearning/quiz/service/KidsGameService.java
package com.elearning.quiz.service;

import com.elearning.quiz.dto.GameCategoryDTO;
import com.elearning.quiz.dto.GameQuestionDTO;
import com.elearning.quiz.dto.KidsGameDTO;
import com.elearning.quiz.model.GameCategory;
import com.elearning.quiz.model.GameQuestion;
import com.elearning.quiz.model.KidsGame;
import com.elearning.quiz.model.QuizStatus;
import com.elearning.quiz.repository.GameCategoryRepository;
import com.elearning.quiz.repository.GameQuestionRepository;
import com.elearning.quiz.repository.KidsGameRepository;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class KidsGameService {

    @Autowired
    private KidsGameRepository gameRepository;

    @Autowired
    private GameCategoryRepository categoryRepository;

    @Autowired
    private GameQuestionRepository questionRepository;

    @PostConstruct
    public void initDefaultCategories() {
        if (categoryRepository.count() == 0) {
            categoryRepository.save(new GameCategory(null, "animals", "🐾", "#FF6B9D", "Learn about cute animals!", null, null));
            categoryRepository.save(new GameCategory(null, "colors", "🎨", "#4ECDC4", "Discover beautiful colors!", null, null));
            categoryRepository.save(new GameCategory(null, "numbers", "🔢", "#FFE66D", "Count and have fun!", null, null));
        }
    }

    // Category Management
    public List<GameCategoryDTO> getAllCategories() {
        return categoryRepository.findAll().stream()
                .map(this::convertCategoryToDTO)
                .collect(Collectors.toList());
    }

    public GameCategoryDTO getCategoryById(Long id) {
        return categoryRepository.findById(id)
                .map(this::convertCategoryToDTO)
                .orElseThrow(() -> new RuntimeException("Category not found"));
    }

    @Transactional
    public GameCategoryDTO createCategory(GameCategoryDTO dto) {
        if (categoryRepository.existsByName(dto.getName())) {
            throw new RuntimeException("Category with this name already exists");
        }
        GameCategory category = new GameCategory();
        category.setName(dto.getName());
        category.setIcon(dto.getIcon());
        category.setColor(dto.getColor());
        category.setDescription(dto.getDescription());
        return convertCategoryToDTO(categoryRepository.save(category));
    }

    @Transactional
    public GameCategoryDTO updateCategory(Long id, GameCategoryDTO dto) {
        GameCategory category = categoryRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Category not found"));
        category.setName(dto.getName());
        category.setIcon(dto.getIcon());
        category.setColor(dto.getColor());
        category.setDescription(dto.getDescription());
        return convertCategoryToDTO(categoryRepository.save(category));
    }

    @Transactional
    public void deleteCategory(Long id) {
        categoryRepository.deleteById(id);
    }

    // Game Management
    public List<KidsGameDTO> getAllGames() {
        return gameRepository.findAll().stream()
                .map(this::convertGameToDTO)
                .collect(Collectors.toList());
    }

    public List<KidsGameDTO> getGamesByCategory(Long categoryId) {
        return gameRepository.findByCategoryId(categoryId).stream()
                .map(this::convertGameToDTO)
                .collect(Collectors.toList());
    }

    public List<KidsGameDTO> getPublishedGames() {
        return gameRepository.findByStatus(QuizStatus.PUBLISHED).stream()
                .map(this::convertGameToDTO)
                .collect(Collectors.toList());
    }

    public List<KidsGameDTO> getPublishedGamesByCategory(Long categoryId) {
        return gameRepository.findByCategoryIdAndStatus(categoryId, QuizStatus.PUBLISHED).stream()
                .map(this::convertGameToDTO)
                .collect(Collectors.toList());
    }

    public KidsGameDTO getGameById(Long id) {
        return gameRepository.findById(id)
                .map(this::convertGameToDTO)
                .orElseThrow(() -> new RuntimeException("Game not found"));
    }

    @Transactional
    public KidsGameDTO createGame(KidsGameDTO dto) {
        GameCategory category = categoryRepository.findById(dto.getCategoryId())
                .orElseThrow(() -> new RuntimeException("Category not found"));
        
        KidsGame game = new KidsGame();
        game.setTitle(dto.getTitle());
        game.setDescription(dto.getDescription());
        game.setCategory(category);
        game.setStatus(dto.getStatus() != null ? dto.getStatus() : QuizStatus.DRAFT);
        
        return convertGameToDTO(gameRepository.save(game));
    }

    @Transactional
    public KidsGameDTO updateGame(Long id, KidsGameDTO dto) {
        KidsGame game = gameRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Game not found"));
        
        if (dto.getCategoryId() != null) {
            GameCategory category = categoryRepository.findById(dto.getCategoryId())
                    .orElseThrow(() -> new RuntimeException("Category not found"));
            game.setCategory(category);
        }
        
        game.setTitle(dto.getTitle());
        game.setDescription(dto.getDescription());
        game.setStatus(dto.getStatus());
        
        return convertGameToDTO(gameRepository.save(game));
    }

    @Transactional
    public void deleteGame(Long id) {
        gameRepository.deleteById(id);
    }

    // Question Management
    public List<GameQuestionDTO> getQuestionsByGame(Long gameId) {
        return questionRepository.findByGameId(gameId).stream()
                .map(this::convertQuestionToDTO)
                .collect(Collectors.toList());
    }

    public GameQuestionDTO getQuestionById(Long id) {
        return questionRepository.findById(id)
                .map(this::convertQuestionToDTO)
                .orElseThrow(() -> new RuntimeException("Question not found"));
    }

    @Transactional
    public GameQuestionDTO createQuestion(GameQuestionDTO dto) {
        KidsGame game = gameRepository.findById(dto.getGameId())
                .orElseThrow(() -> new RuntimeException("Game not found"));
        
        GameQuestion question = new GameQuestion();
        question.setGame(game);
        question.setImageEmoji(dto.getImageEmoji());
        question.setCorrectAnswer(dto.getCorrectAnswer());
        question.setOptions(dto.getOptions());
        question.setPoints(dto.getPoints() != null ? dto.getPoints() : 1);
        
        return convertQuestionToDTO(questionRepository.save(question));
    }

    @Transactional
    public GameQuestionDTO updateQuestion(Long id, GameQuestionDTO dto) {
        GameQuestion question = questionRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Question not found"));
        
        question.setImageEmoji(dto.getImageEmoji());
        question.setCorrectAnswer(dto.getCorrectAnswer());
        question.setOptions(dto.getOptions());
        question.setPoints(dto.getPoints());
        
        return convertQuestionToDTO(questionRepository.save(question));
    }

    @Transactional
    public void deleteQuestion(Long id) {
        questionRepository.deleteById(id);
    }

    // Converters
    private KidsGameDTO convertGameToDTO(KidsGame game) {
        KidsGameDTO dto = new KidsGameDTO();
        dto.setId(game.getId());
        dto.setTitle(game.getTitle());
        dto.setDescription(game.getDescription());
        dto.setCategoryId(game.getCategory().getId());
        dto.setCategoryName(game.getCategory().getName());
        dto.setCategoryIcon(game.getCategory().getIcon());
        dto.setCategoryColor(game.getCategory().getColor());
        dto.setStatus(game.getStatus());
        dto.setQuestionCount(game.getQuestions().size());
        return dto;
    }

    private GameCategoryDTO convertCategoryToDTO(GameCategory category) {
        GameCategoryDTO dto = new GameCategoryDTO();
        dto.setId(category.getId());
        dto.setName(category.getName());
        dto.setIcon(category.getIcon());
        dto.setColor(category.getColor());
        dto.setDescription(category.getDescription());
        return dto;
    }

    private GameQuestionDTO convertQuestionToDTO(GameQuestion question) {
        GameQuestionDTO dto = new GameQuestionDTO();
        dto.setId(question.getId());
        dto.setGameId(question.getGame().getId());
        dto.setImageEmoji(question.getImageEmoji());
        dto.setCorrectAnswer(question.getCorrectAnswer());
        dto.setOptions(question.getOptions());
        dto.setPoints(question.getPoints());
        return dto;
    }
}

File: src/app/kids/services/kids-game.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { KidsGame, GameCategory, GameQuestion } from '../models/kids-game.models';

@Injectable({
  providedIn: 'root'
})
export class KidsGameService {
  private apiUrl = 'http://localhost:8081/api/kids-games';

  constructor(private http: HttpClient) {}

  // Category Management
  getAllCategories(): Observable<GameCategory[]> {
    return this.http.get<GameCategory[]>(`${this.apiUrl}/categories`);
  }

  getCategoryById(id: number): Observable<GameCategory> {
    return this.http.get<GameCategory>(`${this.apiUrl}/categories/${id}`);
  }

  createCategory(category: GameCategory): Observable<GameCategory> {
    return this.http.post<GameCategory>(`${this.apiUrl}/categories`, category);
  }

  updateCategory(id: number, category: GameCategory): Observable<GameCategory> {
    return this.http.put<GameCategory>(`${this.apiUrl}/categories/${id}`, category);
  }

  deleteCategory(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/categories/${id}`);
  }

  // Game Management
  getAllGames(): Observable<KidsGame[]> {
    return this.http.get<KidsGame[]>(this.apiUrl);
  }

  getPublishedGames(): Observable<KidsGame[]> {
    return this.http.get<KidsGame[]>(`${this.apiUrl}/published`);
  }

  getGamesByCategory(categoryId: number): Observable<KidsGame[]> {
    return this.http.get<KidsGame[]>(`${this.apiUrl}/category/${categoryId}`);
  }

  getPublishedGamesByCategory(categoryId: number): Observable<KidsGame[]> {
    return this.http.get<KidsGame[]>(`${this.apiUrl}/published/category/${categoryId}`);
  }

  getGameById(id: number): Observable<KidsGame> {
    return this.http.get<KidsGame>(`${this.apiUrl}/${id}`);
  }

  createGame(game: KidsGame): Observable<KidsGame> {
    return this.http.post<KidsGame>(this.apiUrl, game);
  }

  updateGame(id: number, game: KidsGame): Observable<KidsGame> {
    return this.http.put<KidsGame>(`${this.apiUrl}/${id}`, game);
  }

  deleteGame(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Question Management
  getQuestionsByGame(gameId: number): Observable<GameQuestion[]> {
    return this.http.get<GameQuestion[]>(`${this.apiUrl}/${gameId}/questions`);
  }

  getQuestionById(id: number): Observable<GameQuestion> {
    return this.http.get<GameQuestion>(`${this.apiUrl}/questions/${id}`);
  }

  createQuestion(question: GameQuestion): Observable<GameQuestion> {
    return this.http.post<GameQuestion>(`${this.apiUrl}/questions`, question);
  }

  updateQuestion(id: number, question: GameQuestion): Observable<GameQuestion> {
    return this.http.put<GameQuestion>(`${this.apiUrl}/questions/${id}`, question);
  }

  deleteQuestion(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/questions/${id}`);
  }
}
